torch_np/_ndarray.py

```diff
@@ -0,0 +1,520 @@
+        except Exception:
+            mesg = "only integer scalar arrays can be converted to a scalar index"
+            raise TypeError(mesg)
+
+    def __float__(self):
+        return float(self.tensor)
+
+    def __complex__(self):
+        try:
+            return complex(self.tensor)
+        except ValueError as e:
+            raise TypeError(*e.args)
+
+    def __int__(self):
+        return int(self.tensor)
+
+    # XXX : are single-element ndarrays scalars?
+    # in numpy, only array scalars have the `is_integer` method
+    def is_integer(self):
+        try:
+            result = int(self.tensor) == self.tensor
+        except Exception:
+            result = False
+        return result
+
+    ### sequence ###
+    def __len__(self):
+        return self.tensor.shape[0]
+
+    ### arithmetic ###
+
+    # add, self + other
+    __add__ = __radd__ = _ufuncs.add
+
+    def __iadd__(self, other):
+        return _ufuncs.add(self, other, out=self)
+
+    # sub, self - other
+    __sub__ = _ufuncs.subtract
+
+    # XXX: generate a function just for this? AND other non-commutative ops.
+    def __rsub__(self, other):
+        return _ufuncs.subtract(other, self)
+
+    def __isub__(self, other):
+        return _ufuncs.subtract(self, other, out=self)
+
+    # mul, self * other
+    __mul__ = __rmul__ = _ufuncs.multiply
+
+    def __imul__(self, other):
+        return _ufuncs.multiply(self, other, out=self)
+
+    # div, self / other
+    __truediv__ = _ufuncs.divide
+
+    def __rtruediv__(self, other):
+        return _ufuncs.divide(other, self)
+
+    def __itruediv__(self, other):
+        return _ufuncs.divide(self, other, out=self)
+
+    # floordiv, self // other
+    __floordiv__ = _ufuncs.floor_divide
+
+    def __rfloordiv__(self, other):
+        return _ufuncs.floor_divide(other, self)
+
+    def __ifloordiv__(self, other):
+        return _ufuncs.floor_divide(self, other, out=self)
+
+    __divmod__ = _ufuncs.divmod
+
+    # power, self**exponent
+    __pow__ = __rpow__ = _ufuncs.float_power
+
+    def __rpow__(self, exponent):
+        return _ufuncs.float_power(exponent, self)
+
+    def __ipow__(self, exponent):
+        return _ufuncs.float_power(self, exponent, out=self)
+
+    # remainder, self % other
+    __mod__ = __rmod__ = _ufuncs.remainder
+
+    def __imod__(self, other):
+        return _ufuncs.remainder(self, other, out=self)
+
+    # bitwise ops
+    # and, self & other
+    __and__ = __rand__ = _ufuncs.bitwise_and
+
+    def __iand__(self, other):
+        return _ufuncs.bitwise_and(self, other, out=self)
+
+    # or, self | other
+    __or__ = __ror__ = _ufuncs.bitwise_or
+
+    def __ior__(self, other):
+        return _ufuncs.bitwise_or(self, other, out=self)
+
+    # xor, self ^ other
+    __xor__ = __rxor__ = _ufuncs.bitwise_xor
+
+    def __ixor__(self, other):
+        return _ufuncs.bitwise_xor(self, other, out=self)
+
+    # bit shifts
+    __lshift__ = __rlshift__ = _ufuncs.left_shift
```

Comment:
bit shifts (left and right) are not commutative, so the r-looking variants should be made explicit as well.